File: apps/android/app/src/main/kotlin/app/trackevolution/screens/AnalysisColumn.kt
package app.trackevolution.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.key
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.testTag
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import app.trackevolution.core.ChannelGraphs
import app.trackevolution.core.LapTime
import app.trackevolution.core.TraceSample
// Aliased: `TrackMap` is also the *composable* that draws the map, imported
// below. One is the geometry and one is the picture, and they deliberately share
// the name of the JS module they both come from — so the import says which.
import app.trackevolution.core.TrackMap as TrackMapGeometry
import app.trackevolution.core.model.Session
import app.trackevolution.ui.TESectionHeader
import app.trackevolution.ui.charts.ChannelHit
import app.trackevolution.ui.charts.LapChannelChart
import app.trackevolution.ui.charts.LimitLegend
import app.trackevolution.ui.charts.TrackMap
import app.trackevolution.ui.theme.TrackTheme

/**
 * The event page's right-hand column at expanded width (spec: NS-34 ticket 3).
 *
 * The selected session's **track map above its channel panel**, so map and charts
 * are in one eyeline. On a phone the panel lives inside the session card, a long
 * scroll away from the trace at the top of the page; a window wide enough to show
 * both at once is what makes the pairing worth having, and this is the container
 * that provides it.
 *
 * Its own scroll, deliberately: the page beside it is long and this is not, and
 * tying them together would mean scrolling past the charts to reach the sessions.
 */
@Composable
fun AnalysisColumn(
    sessions: List<Session>,
    selectedSessionId: Int?,
    modifier: Modifier = Modifier,
) {
    val colors = TrackTheme.colors
    val selected = sessions.firstOrNull { it.id == selectedSessionId }

    // Where the panel is pointing, if anywhere — the friction circle's tapped
    // sample. Held here rather than in the panel because the *map* is what
    // answers it, and the map is this column's, not the panel's.
    var hit by remember(selectedSessionId) { mutableStateOf<ChannelHit?>(null) }

    Column(
        modifier = modifier
            .fillMaxSize()
            .background(colors.bgPage)
            .verticalScroll(rememberScrollState())
            .padding(16.dp)
            .semantics { testTag = "analysisColumn" },
        verticalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        if (selected == null) {
            Empty(anyChannels = sessions.any { it.channels != null })
            return@Column
        }

        val trace = selected.trace
        if (!trace.isNullOrEmpty()) {
            TESectionHeader("Best lap trace", detail = "brighter is faster")
            Text(
                LapTime.fmtMs(selected.bestLapMs),
                style = TrackTheme.typography.lapTime,
                color = colors.textStrong,
            )
            // The stored trace is `TracePoint`s off the wire; the chart draws
            // `TraceSample`s, as the page's own trace section does.
            val samples = trace.map { TraceSample(x = it.x, y = it.y, v = it.v) }
            val markers = limitMarkers(selected, samples)
            TrackMap(
                trace = samples,
                markers = markers,
                highlight = ringedIndex(selected, hit),
            )
            LimitLegend(markers)
        }

        val channels = selected.channels
        if (channels != null) {
            TESectionHeader(selected.label ?: "Channel graphs")
            // Keyed by session id: `lit` holds *channel-lap indexes*, which mean
            // different laps in a different session, so carrying them across
            // would light the wrong ones. Remembering tab and lit state per
            // session — which NS-34 asks for — needs the panel to hand that state
            // back out, and is deferred rather than faked; see the note on #217.
            key(selected.id) {
                LapChannelChart(
                    channels = channels,
                    laps = selected.laps,
                    modifier = Modifier.fillMaxWidth(),
                    onHit = { hit = it },
                )
            }
        }
    }
}

@Composable
private fun Empty(anyChannels: Boolean) {
    val colors = TrackTheme.colors
    Column(
        modifier = Modifier.fillMaxWidth().padding(top = 48.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        Text(
            if (anyChannels) "Pick a session" else "No channel data yet",
            style = TrackTheme.typography.h3,
            color = colors.textStrong,
        )
        Text(
            if (anyChannels) {
                "Choose a session's channel graphs to see them here."
            } else {
                "Import a video or record laps on the phone, and the channels land here."
            },
            style = TrackTheme.typography.sm,
            color = colors.textMuted,
            textAlign = TextAlign.Center,
            modifier = Modifier.widthIn(max = 300.dp),
        )
    }
}

/**
 * Where to ring the map for the current hit, or null for "don't".
 *
 * The stored trace is **one lap** — the session's best — so a sample from any
 * other lap has no place on it, and ringing one anyway would put the mark where
 * that lap never was. Same constraint the limit marks work under
 * ([Limits.limitMarkers]), and the same rule the web states in `bindBalance`: a
 * hit with no lap of its own (`chIdx` null) is a place every lap shares, so it
 * rings whichever lap the trace is.
 */
private fun ringedIndex(session: Session, hit: ChannelHit?): Int? {
    if (hit == null) return null
    val trace = session.trace ?: return null
    if (hit.chIdx != null && hit.chIdx != tracedChannelIndex(session)) return null
    return TrackMapGeometry.traceIndexAtFraction(trace, hit.frac)
}

/**
 * The channel-lap the stored trace was drawn from: the session's fastest lap that
 * has channel data, matched the way the panel matches them.
 */
private fun tracedChannelIndex(session: Session): Int? {
    val channels = session.channels?.takeIf { it.laps.isNotEmpty() } ?: return null
    return ChannelGraphs.matchLapsToChannels(session.laps, channels.laps)
        .filter { it.hasChannels }
        .minByOrNull { it.lap.timeMs }
        ?.chIdx
}

File: apps/ios/App/Screens/AnalysisColumn.swift
import SwiftUI
import TrackEvolutionKit

/// The event page's right-hand column at expanded width (spec: NS-34 ticket 3).
///
/// The best-lap **track map above the channel panel**, for the session selected in
/// the page beside it, so that map and charts are in one eyeline. That ordering is
/// the point of the column rather than a preference: on a phone the panel is a
/// sheet *over* the map, which is why the web's hover behaviour — mark the
/// distance across the charts, ring the place on the map — has never been worth
/// porting. A window wide enough to show both at once is what makes it worth
/// having, and this is the container that provides it.
///
/// It has **its own `ScrollView`** and is a sibling of the page's `List`, never a
/// row inside it. `LapChannelChart` documents why: a chart of that many marks
/// inside a `List` row never settles.
struct AnalysisColumn: View {
    let detail: EventDetail
    @Binding var selectedSessionId: Int?

    /// Where the panel is pointing, if anywhere — the friction circle's tapped
    /// sample. Held here rather than in the panel because the *map* is what
    /// answers it, and the map is this column's, not the panel's.
    @State private var hit: ChannelHit?

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: TESpacing.gridGap) {
                if let session = selected {
                    traceCard(session)
                    panel(session)
                } else {
                    empty
                }
            }
            .padding(TESpacing.pageGutter)
            .frame(maxWidth: .infinity, alignment: .leading)
        }
        .background(Color(.bgPage))
        .accessibilityIdentifier("analysisColumn")
    }

    /// The selected session, falling back to nothing rather than to an arbitrary
    /// one: a column showing a session the page has not highlighted would be
    /// answering a question nobody asked.
    private var selected: Session? {
        guard let id = selectedSessionId else { return nil }
        return detail.sessions.first { $0.id == id }
    }

    /// The racing line of the selected session — its own trace, not the event's
    /// best, because the charts under it are this session's.
    @ViewBuilder
    private func traceCard(_ session: Session) -> some View {
        if let trace = session.trace, !trace.isEmpty {
            TECard {
                VStack(alignment: .leading, spacing: 10) {
                    HStack {
                        Text("Brighter is faster")
                            .teStyle(.xs)
                            .foregroundStyle(Color(.textFaint))
                        Spacer()
                        TETime(ms: session.bestLapMs, emphasized: true)
                    }
                    let markers = EventScreen.limitMarkers(session)
                    TrackMapView(
                        trace: trace,
                        markers: markers,
                        highlight: ringedIndex(session, trace: trace)
                    )
                    .frame(height: 240)
                    LimitLegend(markers: markers)
                }
            }
        }
    }

    @ViewBuilder
    private func panel(_ session: Session) -> some View {
        if let channels = session.channels, !ChannelGraphs.presentChannels(channels).isEmpty {
            VStack(alignment: .leading, spacing: 6) {
                Text(session.label ?? "Channel graphs")
                    .teStyle(.h3)
                    .foregroundStyle(Color(.textStrong))
                LapChannelChart(channels: channels, laps: session.laps, onHit: { hit = $0 })
                    // Keyed by session: `lit` holds *channel-lap indexes*, which
                    // mean different laps in a different session, so carrying them
                    // across would light the wrong ones. Remembering tab and lit
                    // state per session id — which NS-34 asks for — needs the
                    // panel to hand that state back out, and is deferred rather
                    // than faked; see the note on #217.
                    .id(session.id)
            }
        }
    }

    private var empty: some View {
        VStack(spacing: 10) {
            Image(systemName: "chart.xyaxis.line")
                .teStyle(.h1)
                .foregroundStyle(Color(.textFaint))
            Text(hasAnyChannels ? "Pick a session" : "No channel data yet")
                .teStyle(.h3)
                .foregroundStyle(Color(.textStrong))
            Text(
                hasAnyChannels
                    ? "Choose a session's channel graphs to see them here."
                    : "Import a video or record laps on the phone, and the channels land here."
            )
            .teStyle(.sm)
            .foregroundStyle(Color(.textMuted))
            .multilineTextAlignment(.center)
        }
        .frame(maxWidth: .infinity)
        .padding(.top, 60)
    }

    /// Where to ring the map for the current hit, or nil for "don't".
    ///
    /// The stored trace is **one lap** — the session's best — so a sample from
    /// any other lap has no place on it, and ringing one anyway would put the
    /// mark where that lap never was. Same constraint the limit marks work under
    /// (`Limits.limitMarkers`), and the same rule the web states in `bindBalance`:
    /// a hit with no lap of its own (`chIdx` nil) is a place every lap shares, so
    /// it rings whichever lap the trace is.
    private func ringedIndex(_ session: Session, trace: [TracePoint]) -> Int? {
        guard let hit else { return nil }
        if let chIdx = hit.chIdx, chIdx != Self.tracedChannelIndex(session) { return nil }
        return TrackMap.traceIndexAtFraction(trace, hit.frac)
    }

    /// The channel-lap the stored trace was drawn from: the session's fastest lap
    /// that has channel data, matched the way the panel matches them.
    private static func tracedChannelIndex(_ session: Session) -> Int? {
        guard let channels = session.channels, !channels.laps.isEmpty else { return nil }
        return ChannelGraphs.matchLapsToChannels(session.laps, channels.laps)
            .filter(\.hasChannels)
            .min { $0.lap.timeMs < $1.lap.timeMs }?
            .chIdx
    }

    private var hasAnyChannels: Bool {
        detail.sessions.contains { session in
            session.channels.map { !ChannelGraphs.presentChannels($0).isEmpty } ?? false
        }
    }
}
